### ISSUE TITLE
Fixing memory leak reports with VLD when linking statically

### ISSUE THREAD

**Issue by dmpriso** [nats-io/cnats#68]
As discussed on slack, VLD reports several small leaks when linking cnats statically. This is because thread-local and global allocations are actually released during cnats `DllMain()` function (or it's unix equivalent). However, `DllMain()` will not be called obviously when static linking (and may also introduce symbol conflicts when creating a DLL which statically links to cnats).
While the allocations are actually uncritical, since they are small and not growing, they may make hunting for actual leaks harder (in the end, everyone wants a clean memory report after debugging)

I've created a fix for that already but one small thing is still left to discuss.

First of all, I've created a `NATS_STATIC` symbol and avoid creating of `DllMain()` when defined. Then, I've created a `NATS_EXAMPLES_STATIC_LINKING` option to cmake for testing.

For the actual fix, I've created and implemented two new functions:

```C
/** \brief Per-Thread cleanup function for static library
*
* When linking to the static library, this should be called at the end of
* each thread where NATS objects have been created or used.
*
* \note This has no effect when called from the shared library.
* nats_Close() implicitly calls that.
*/
NATS_EXTERN void
nats_ReleaseThreadMemory(void);

/** \brief Global cleanup function for static library
*
* When linking to the static library, this should be called at the end of
* the application in order to perform library cleanup.
* IMPORTANT The library may not be re-used again after calling this.
*
* \note This has no effect when called from the shared library.
*/
NATS_EXTERN void
nats_ReleaseStaticLibrary(void);
```

`nats_ReleaseThreadMemory()` is already called by `nats_Close()`, so for many actual implementations it shouldn't be necessary to call that.

In my first try, I also called `nats_ReleaseStaticLibrary()` from `nats_Close()`. However this *may* break existing code: Until now, it was possible to "reuse" the library after `nats_Close()` has been called. However, I assume the intention of calling `nats_Close()` is to perform some final cleanup. If that is the case, we may also call `nats_ReleaseStaticLibraray()` from `nats_Close()`, making the whole static library cleanup automatic (except for multi-threaded applications where `nats_ReleaseThreadMemory()` is still required to be called for obvious reasons).

Background: `nats_ReleaseStaticLibrary()` performs the cleanup which a shared lib does during `DllMain(DLL_PROCESS_DETACH)`, which means deleting the mutex. However, the mutex cannot be reinitialized cause for doing so `gInitOnce` must be reset (which cannot be done in clean code; we could of course hack this)

I've got my code ready for creating a PR but will wait for feedback on this issue before filing.

**Comment by dmpriso** [comment]
I also noticed that cnats creates it's own threads for certain tasks and built-in the per-thread cleanup code there too.

**Comment by kozlovic** [comment]
@dmpriso Thanks for the report!

As you have seen, the library creates its own threads. I don't think it is advisable to ask the user to call `nats_ReleaseThreadMemory()`. Instead, I may be able to do the thread cleanup with use of hooks in `_threadStart`.

As for freeing the global library lock and thread local keys, then this would have to be done in a new API as you suggest. I don't want to do that in existing `nats_Close()` for the reasons you mentioned.

Let me play a little bit and see if it is possible to move the thread cleanup directly in `_threadStart` (removing the need to be in `DllMain`) and will get back to you.

If your PR is ready, you can always submit, with the understanding that we may not end-up merging it in its entirety if we go another direction for thread cleanup.

**Comment by dmpriso** [comment]
Well, if cnats only allocates TLS in cnats-created threads then letting the user call `nats_ReleaseThreadMemory()` is not required actually.

I already have put the cleanup code into the thread creation stubs, see the PR I've just filed. (This however still contains the `nats_ReleaseThreadMemory()` API)

**Comment by dmpriso** [comment]
I found that `target_compile_definitions` seems to be unsupported by the used cmake version? It should be available since 3.0.2 so quite long. Hmm. Used that to tell the code that it's a static library. I think this is required for the process cleanup at least.
I can have a look and look for some alternative tomorrow, I'm at GMT+1 and I'm not at work any more.
